### ISSUE TITLE
Use cases and shortcomings for the Responsive component

### ISSUE THREAD

**Issue by theetrain** [telusdigital/tds#465]
## Why

Digital Platform developers traditionally wrote CSS media queries for all their responsive views. It wasn't possible to write unit tests for responsive design, plus writing CSS proved to be time-consuming in general. In response, the `<Responsive />` component was made to tackle these issues.

## What

The `<Responsive />` component was included in TDS as a replacement for CSS media queries. It wraps the [react-media](https://github.com/ReactTraining/react-media/) community component. The key benefits are:

- developers can write unit tests for responsive components
- responsive design is visible within JSX, which maintains context and reduces cognitive strain
- by wrapping other components, you can have responsive JS such as variables that show a different value on desktop and mobile viewports
- performance improvements by letting JS decide which component view to render. This way you can have two or more DOM elements conditionally render rather than having them be present and depend on CSS

There are some known shortcomings with the `<Responsive />` component:

- the starter kit is missing user agent sniffing functionality, which causes issues with server-side rendering. A fix is underway #461 

## Discuss

Has the `<Responsive />` component fallen short for you? Do you have some particular pain points that we can reconcile?

**Comment by xingbo828** [comment]
1. On server side, instead of rendering both desktop and mobile version. I believe render nothing is better. This is the default behaviour of react-media. Somehow it is altered in the `Responsive` component.
2. I don't believe relying on the user-agent is a great idea for the following 2 reasons:
   - TDS should be a stand-alone lib instead of relying on the hosting app to provide additional informations
   - I think user-agent can only gives info about browser type and version, but not the width of the browser. (iPad mobile Safari portraits vs landscape )

**Comment by ryanoglesby08** [comment]
Re: render nothing on server and default behaviour of react-media. -- `Responsive` component does not alter the behaviour of react-media.

Looking at react-media source, it uses a state property of `matches` to determine whether or not to render when there is no `window`. It is set to `true` by default so that it always matches. See here: https://github.com/ReactTraining/react-media/blob/master/modules/Media.js#L21-L27.

The `Responsive` component allows the use of the `defaultMatches` prop of react-media so that consumers can control this behaviour.

We _could_ set `defaultMatches` to false in the `Responsive` component... but not sure if that is the way to go?

**Comment by ryanoglesby08** [comment]
Re: user agent

Totally agree that TDS should be standalone! That is why we have not built any user agent stuff into the component and want to rely on the consuming app to build additional functionality on top of the low-level, raw TDS components. We want to keep global app level concerns out of TDS.

And yes, user agent only allows you to make an educated guess about the client's device characteristics, which might be "good enough"?

**Comment by lzcabrera** [comment]
> We could set defaultMatches to false in the Responsive component... but not sure if that is the way to go?

I'm afraid that setting `defaultMatches` to false by default is not the way to go due to the SEO implications (and defeats the purpose of having an isomorphic starter kit).

**Comment by githubjosh** [comment]
@lzcabrera you beat me to it. Yes! 

I might suggest a mobile first approach to rendering: ie. when in doubt optimize for mobile. 

But all important content should be server rendered. That is essential

**Comment by ryanoglesby08** [comment]
See #461.

For reference, we are likely implementing user-agent detection so that the server can intelligently decide what to render depending on the client. This will have the added benefit of potentially small payloads on initial render vs plain media queries.

**Comment by githubjosh** [comment]
^^ I feel like user agent detection was a rabbit hole that mobile first design was intended to resolve.

Is this component necessary? I feel like we're potentially adding to the problem rather than removing it. 

What will bots see?

**Comment by githubjosh** [comment]
Also without doing any detailed analysis a reduction in html on the server side render is unlikely to result in a sizable performance gain relative to the size of your js bundle.

Any idea if this component is in wide use? I might suggest deprecating it quickly and revisiting its value. Just not sure if this is currently solving a problem.

**Comment by ryanoglesby08** [comment]
@githubjosh Thanks for your comments and both issues! Good points.

> Is this component necessary? I feel like we're potentially adding to the problem rather than removing it.

> I think it may be worth revisiting the value of this component given the noted issues. Again, we did not have this issue prior. What problem are we trying to solve here?

I think its best to give the full picture here:

### What is the Responsive component?

A component interface to media queries through JavaScript.

### What's wrong with media queries in CSS?

CSS media queries are essentially conditional logic embedded in CSS, which (among other things) contributes to making CSS codebases difficult to maintain and reason about over time.

CSS media queries suffer from:
* the corresponding markup often must contain multiple elements for all queries
* having CSS logic that is separate from markup and JS logic is hard to conceptualize
* you cannot easily test CSS media queries
* it is difficult to optimize your CSS bundles for various clients
* CSS media queries are limited in power to only styling

### What does the Responsive component provide?

* you can render a component or not based on the result of the query
* you can render one component or another based on the query
* you can write unit tests for media query behaviour
* the component is more likely to be able to easily expand to other platforms (think react-native-web)
* because it's "just a component", consumers can easily compose it to build higher order components without reimplementing things (think responsive grid)
* with a little more smarts, we can optimize critical path styles and markup for faster load times

Looking at the bigger picture, because TD has embraced ["component architecture"](https://github.com/telusdigital/reference-architecture/blob/master/development/react.md), providing a JS-first, component-oriented interface to the design system fits nicely. By standardizing the interface across TDS to just components, we can lessen the cognitive load of using the system. If you need something from TDS, it's a component.

### What's the catch?

Because Responsive uses `window.matchMedia`, when rendering on the server the component does not know which styles of components to render. We are seeing a brief flash of content in the browser when the server rendered markup/styles differs from the client side version.

We believe this is due to a delay in JS execution on the client. The pre-rendered app is mounted, styles are applied and the first paint happens, then the `componentDidMount` is called, triggering the media query, which causes the visible shift in style.

### The hypothesis

We can use a contained bit of user agent detection (which seems to be the way others are going too in this situation) to control how the Responsive component behaves on the server, so that is can make a good enough guess at the client and know how to render itself before returning to the client. This should fix the flash of content, while also giving a slight optimization to the initial payload.

### Conclusion

I believe that all of this creates a pretty compelling story for the Responsive component. This component in particular opens up a lot of future possibilities and enables a more powerful and expressive way of writing responsive Web applications.

**Comment by ryanoglesby08** [comment]
@githubjosh 

> I feel like user agent detection was a rabbit hole that mobile first design was intended to resolve.

Agreed. We don't intend for this to in any way change how design is done, or replace mobile-first design.

If we do go this route, we could keep the user agent detection contained in scope, and not available for nefarious use.

**Comment by ryanoglesby08** [comment]
@githubjosh 

> What will bots see?

Good question! We would have to keep that in mind. I would propose no change in current behavior, bots are served the default rendering of the app. No special behaviour should be applied to bots is the best thing to do I believe. Serve them just as you would a human in a web browser.

**Comment by ryanoglesby08** [comment]
Wanted to update here...

Been having lots of great conversations with others on this. We are currently re-evaluating the implementation of the Responsive component, and re-considering the advice of using user agent detection to help alleviate the server side rendering issues.

Thanks for your patience and feedback. Stay tuned for more details.

**Comment by deluan** [comment]
>4. Refactor TDS components to use the Sass mixins as needed (e.g. Button, Box, DisplayHeading, Heading, ExpandCollapse) (non-breaking patches)

This item is causing a bug in Annual Reports animation (it is triggering the animation twice), and our release date is coming soon (less than two weeks). See the current behaviour here (same user/pass as wcstage):
https://wcpreview.telus.com/en/bc/about/investors/annualreports/2017

Is it possible to release this particular fix as a patch sooner than later? Not having this will impact our designs and we won't have time to pivot to a new one in time for the release.

**Comment by theetrain** [comment]
Patched in this release: https://github.com/telusdigital/tds/releases/tag/%40tds%2Fcore-responsive%401.1.0
